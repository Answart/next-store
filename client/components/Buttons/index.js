export { AddToCart } from './AddToCart';
export { CheckoutCart } from './CheckoutCart';
export { DeleteProduct } from './DeleteProduct';
export { DeleteProductVariant } from './DeleteProductVariant';
export { Logout } from './Logout';
export { RemoveFromCart } from './RemoveFromCart';
export { RequestPasswordReset } from './RequestPasswordReset';
export { ToggleCart } from './ToggleCart';
export { UpdateCartItem } from './UpdateCartItem';
export { UpdatePermissions } from './UpdatePermissions';
