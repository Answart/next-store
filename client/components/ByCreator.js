import PropTypes from 'prop-types';
import Link from 'next/link';


const ByCreator = ({ name, online }) => (
  <div className="prdct-creator prdct-padding">
    By
    <Link href={{
      pathname: "/shop",
      query: { name, online }
    }}><a className="undrln-btn">
      {name}
    </a></Link>
  </div>
);

ByCreator.propTypes = {
  name: PropTypes.string.isRequired,
  online: PropTypes.bool.isRequired
};


export default ByCreator;
