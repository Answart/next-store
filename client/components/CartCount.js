import React from 'react';
import PropTypes from 'prop-types';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import { StyledCartCountAnimation, StyledCartCountDot } from './styles/CartStyles';


const CartCount = ({ count }) => (
  <StyledCartCountAnimation>
    <TransitionGroup>
      <CSSTransition
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
        unmountOnExit
      >
        <StyledCartCountDot>
          <div className={!count ? '' : 'active-count'}>
            {count}
          </div>
        </StyledCartCountDot>
      </CSSTransition>
    </TransitionGroup>
  </StyledCartCountAnimation>
);

CartCount.propTypes = {
  count: PropTypes.number.isRequired,
};


export default CartCount;
