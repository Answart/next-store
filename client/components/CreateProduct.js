import React, { Component } from 'react';
import Link from 'next/link';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import StyledForm from './styles/FormStyles';
import StyledProduct, { StyledProductVariant } from './styles/ProductStyles';
import { user } from '../lib/dummyData';
import { CREATE_PRODUCT_MUTATION } from '../graphql';


const departments = ['Tops', 'Bottoms', 'Shoes', 'Outwear', 'Accessories', 'Decor', 'Wedding'];
const categoriesByDept = {
  'Tops': ['Blouses', 'Cartigans', 'Casual', 'Sweaters', 'Sweatshirts'],
  'Bottoms': ['Skirts', 'Pants', 'Jeans', 'Leggings', 'Shorts', 'Intimates'],
  'Shoes': ['Heels', 'Flats', 'Sandals', 'Wedges', 'Boots', 'Booties', 'Sneakers'],
  'Outwear': ['Jackets', 'Blazers', 'Coats', 'Cartigans'],
  'Accessories': ['Jewelry', 'Bags', 'Wallets', 'Belts', 'Sunglasses', 'Makeup', 'Hair & Hats'],
  'Decor': ['Home', 'Kitchen', 'Office', 'Lighting'],
  'Wedding': ['Dresses', 'Shoes', 'Gifts', 'Accessories']
};

class CreateProduct extends Component {
  state = {
    department: 'Tops',
    title: '',
    description: '',
    image: '',
    category: '',
    brand: ''
  };
  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    if (name === 'department') {
      this.setState({
        category: '',
        [name]: val
      });
    } else {
      this.setState({ [name]: val });
    }
  };
  uploadFile = async e => {
    const files = e.target.files;

    // Upload file and return url;
    const image = files.length
      ? files[0].name
      : '';

    this.setState({ image });
  };
  render() {
    const {
      title,
      department,
      description,
      image,
      category,
      brand
    } = this.state;
    let categories = department
      ? categoriesByDept[department]
      : categoriesByDept['Tops'];
    return (
      <Mutation
        mutation={CREATE_PRODUCT_MUTATION}
        variables={this.state}
      >
        {(createProduct, { loading, error }) => (
          <StyledForm
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createProduct();
              Router.push({
                pathname: '/product/update',
                query: { id: res.data.createProduct.id },
              });
            }}
          >

            <fieldset disabled={false} aria-busy={false}>
              <StyledProduct>
                <div className="buy-prdct-imgs">
                  {image && (
                    <img width="200" src={image} alt="Upload Preview" />
                  )}
                </div>

                <div className="buy-prdct-content">
                  <label htmlFor="title" className="buy-prdct-title buy-prdct-padding">
                    Title:
                    <input
                      type="text"
                      id="title"
                      name="title"
                      placeholder="Title"
                      value={title}
                      onChange={this.handleChange}
                      required
                    />
                  </label>

                  <div className="buy-prdct-creator buy-prdct-padding">
                    By <strong>{user.name}</strong>
                  </div>

                  <label htmlFor="department">
                    Department:
                    <select
                      id="department"
                      name="department"
                      form="carform"
                      placeholder="Department i.e. Shoes"
                      value={department}
                      onChange={this.handleChange}
                      required
                    >
                      {departments.map(dept => <option key={dept} value={dept}>{dept}</option>)}
                    </select>
                  </label>

                  <label htmlFor="category">
                    Category:
                    <select
                      id="category"
                      name="category"
                      form="carform"
                      value={category}
                      onChange={this.handleChange}
                    >
                      <option key={0} value={''}></option>
                      {categories.map(ctgry => <option key={ctgry} value={ctgry}>{ctgry}</option>)}
                    </select>
                  </label>

                  <label htmlFor="description" className="buy-prdct-desc">
                    Description:
                    <textarea
                      id="description"
                      name="description"
                      placeholder="Enter A Description"
                      value={description}
                      onChange={this.handleChange}
                      required
                    />
                  </label>

                  <label htmlFor="brand" className="buy-prdct-brand">
                    Brand:
                    <input
                      type="text"
                      id="brand"
                      name="brand"
                      placeholder="Brand"
                      value={brand}
                      onChange={this.handleChange}
                    />
                  </label>

                  <label htmlFor="img">
                    Image:
                    <input
                      type="file"
                      id="img"
                      name="img"
                      placeholder="Upload an image"
                      onChange={this.uploadFile}
                      required
                    />
                  </label>
                </div>
              </StyledProduct>

              <div className="form-actions buy-prdct-padding">
                <button
                  type="submit"
                  className="buy-prdct-btn"
                >Create</button>
              </div>
              <div className="form-actions buy-prdct-padding">
                <button type="cancel">
                  <Link href="/">
                    <a>Cancel</a>
                  </Link>
                </button>
              </div>
            </fieldset>
          </StyledForm>
        )}
      </Mutation>
    );
  }
}

export default CreateProduct;
