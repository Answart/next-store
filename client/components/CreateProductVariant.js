import React, { Component } from 'react';
import StyledForm from './styles/FormStyles';
import StyledProduct from './styles/ProductStyles';


const colors = ['Red', 'Orange', 'Yellow', 'Green', 'Blue', 'Purple', 'Pink', 'Brown', 'Black', 'Silver', 'White', 'Gold', 'Multi'];
const sizes = [
  'XXS', 'XS', 'S', 'M', 'L',
  'XL', '1X', '2X', '3X', '4X',
  '0', '2', '4', '6', '8',
  '10', '12', '14', '16', '18'
];

class CreateProductVariant extends Component {
  state = {
    price: 0,
    quantity: 1,
    color: '',
    size: '',
    sale: false,
    salePrice: 0
  };
  handleChange = e => {
    const { name, type, value, checked } = e.target;
    let val = value;
    if (type === 'number') val = parseFloat(value);
    if (type === 'checkbox') val = checked;
    if (name === 'sale') {
      this.setState({
        salePrice: 0,
        [name]: val
      });
    } else {
      this.setState({ [name]: val });
    }
  };
  render() {
    const {
      price,
      quantity,
      color,
      size,
      sale,
      salePrice
    } = this.state;
    return (
      <StyledForm
        data-test="form"
        onSubmit={async e => {
          e.preventDefault();
          console.log('onSubmit', this.state);
        }}
      >

        <fieldset disabled={false} aria-busy={false}>
          <StyledProduct>
            <div className="buy-prdct-content">

              <label htmlFor="size">
                Size:
                <select
                  id="size"
                  name="size"
                  value={size}
                  onChange={this.handleChange}
                >
                  <option key={0} value={''}></option>
                  {sizes.map(sz => <option key={sz} value={sz}>{sz}</option>)}
                </select>
              </label>

              <label htmlFor="color">
                Color:
                <select
                  id="color"
                  name="color"
                  placeholder=""
                  value={color}
                  onChange={this.handleChange}
                >
                  <option key={0} value={''}></option>
                  {colors.map(clr => <option key={clr} value={clr}>{clr}</option>)}
                </select>
              </label>

              <label htmlFor="quantity" className="buy-prdct-brand">
                Quantity:
                <input
                  type="number"
                  id="quantity"
                  name="quantity"
                  placeholder="Quantity"
                  min="1"
                  value={quantity}
                  onChange={this.handleChange}
                  required
                />
              </label>

              <label htmlFor="price" className="buy-prdct-title buy-prdct-padding">
                Price per Item:
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="42.99"
                  min="1"
                  value={price}
                  onChange={this.handleChange}
                  required
                />
              </label>

              <div>
                <input
                  type="checkbox"
                  id="sale"
                  name="sale"
                  value={sale}
                  onChange={this.handleChange}
                />
                <label htmlFor="sale" className="buy-prdct-padding">
                  Is this selection on sale?
                </label>
              </div>

              {sale && (
                <label htmlFor="salePrice" className="buy-prdct-title buy-prdct-padding">
                  Sale Price per Item:
                  <input
                    type="number"
                    id="salePrice"
                    name="salePrice"
                    placeholder="22.50"
                    min="0"
                    disabled={!sale}
                    value={salePrice}
                    onChange={this.handleChange}
                  />
                </label>
              )}

            </div>
          </StyledProduct>

          <div className="form-actions buy-prdct-padding">
            <button type="submit" className="buy-prdct-btn">Add Selection</button>
          </div>

        </fieldset>
      </StyledForm>
    );
  }
}

export default CreateProductVariant;
