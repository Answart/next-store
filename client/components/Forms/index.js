export { SignupForm } from './SignupForm.js';
export { SigninForm } from './SigninForm.js';
export { ResetPasswordForm } from './ResetPasswordForm.js';
export { CreateProductForm } from './CreateProductForm.js';
export { CreateProductVariantForm } from './CreateProductVariantForm.js';
export { UpdateProductForm } from './UpdateProductForm.js';
export { UpdateProductVariantForm } from './UpdateProductVariantForm.js';
