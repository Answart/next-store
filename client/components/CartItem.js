import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import PriceTag from './PriceTag';
import SvgIcon from './SvgIcon';
import { RemoveFromCart, UpdateCartItem } from './Buttons';
import { formatMoney, capWord } from '../lib/utils';


const CartItem = ({ id, quantity, variant }) => {
  if (!variant)
    return (
      <tr>
        <td className="cart-item-details">
          <p>This Product has been removed</p>
        </td>

        <td className="cart-item-quantity">
          <div className="cart-item-quantity-remove">
            <RemoveFromCart id={id} />
          </div>
        </td>
      </tr>
    );
  const title = variant.product.title;
  const price = variant.sale
    ? variant.salePrice
    : variant.price;
  return (
    <tr>
      <td className="cart-item-details">
        <div className='cart-item-image'>
          <Link href={{
            pathname: '/shop',
            query: { id: variant.product.id }
          }}>
            <a>
              <img width="100" height="120"
                src={variant.image.image_url}
                alt={title}
                title={`Go to: ${title}`}
              />
            </a>
          </Link>
        </div>

        <div className='cart-item-meta'>
          <div className="cart-item-detail">
            <h4>{title}</h4>
          </div>

          <div className="cart-item-detail">
            <PriceTag
              price={variant.price}
              sale={variant.sale}
              salePrice={variant.salePrice}
            />
          </div>

          {variant.size && (
            <div className="cart-item-detail">
              Size: {variant.size}
            </div>
          )}

          {variant.color && (
            <div className="cart-item-detail">
              Color: {capWord(variant.color)}
            </div>
          )}
        </div>
      </td>

      <td className="cart-item-quantity">
        <div className="cart-item-quantity-actions">
          <UpdateCartItem
            id={variant.id}
            quantity={quantity - 1}
            disabled={(quantity - 1) <= 0}
          >
            <SvgIcon width={7} name='minus' color='darkBlue' />
          </UpdateCartItem>

          {quantity}

          <UpdateCartItem
            id={variant.id}
            quantity={quantity + 1}
            disabled={(quantity + 1) > variant.quantity}
          >
            <SvgIcon width={7} name='plus' color='darkBlue' />
          </UpdateCartItem>

          <div className="cart-item-quantity-remove">
            <RemoveFromCart id={id} />
          </div>
        </div>
      </td>

      <td className="cart-item-total-price">
        {formatMoney(price * quantity)}
      </td>
    </tr>
  );
};

CartItem.propTypes = {
  id: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  variant: PropTypes.shape({
    price: PropTypes.number.isRequired,
    quantity: PropTypes.number.isRequired,
    size: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
    sale: PropTypes.bool.isRequired,
    salePrice: PropTypes.number.isRequired,
    image: PropTypes.shape({
      id: PropTypes.string.isRequired,
      image_url: PropTypes.string.isRequired
    }),
    product: PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired
    }).isRequired
  })
};


export default CartItem;
