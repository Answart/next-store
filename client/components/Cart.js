import { Query } from 'react-apollo';
import { StyledCart } from './styles/CartStyles';
import { StyledCartTable, StyledTotalsTable } from './styles/TableStyles';
import User from './User';
import CartItem from './CartItem';
import NotFound from './NotFound';
import SvgIcon from './SvgIcon';
import { ToggleCart, CheckoutCart } from './Buttons';
import { formatMoney, getCartTotals } from '../lib/utils';
import { LOCAL_CARTOPEN_QUERY } from '../graphql';


const Cart = ({ me }) => (
  <Query query={LOCAL_CARTOPEN_QUERY}>
    {({ data }) => {
      const myCart = (!!me && !!me.cart)
        ? me.cart
        : [];
      const imgUrl = (!!myCart.length && !!myCart[0].variant)
        ? myCart[0].variant.image.image_url
        : "";
      const { totalQuantity, totalShipping, totalSalesTax, subTotal } = getCartTotals(myCart);
      const totalAmount = (subTotal + totalShipping + totalSalesTax);
      return (
        <StyledCart open={!!data ? data.cartOpen : false}>
          <header>
            <div className="cart-title">
              <ToggleCart>
                <SvgIcon width={10} name='ex' />
              </ToggleCart>

              <h2>Your Cart</h2>
            </div>
          </header>

          <div className="cart-container">
            <StyledCartTable>
              <thead>
                <tr>
                  <th>Product</th>
                  <th>Quantity</th>
                  <th>Total Price</th>
                </tr>
              </thead>

              <tbody>
                {!me && (
                  <tr>
                    <td className="cart-page">
                      <NotFound status={400} />
                    </td>
                  </tr>
                )}
                {!!me && myCart.map(cartItem =>
                  <CartItem key={cartItem.id} id={cartItem.id}
                    quantity={cartItem.quantity}
                    variant={cartItem.variant}
                  />
                )}
              </tbody>
            </StyledCartTable>
          </div>

          <footer>
            <div>
              <StyledTotalsTable>
                <tbody>
                  <tr>
                    <td>Subtotal</td>
                    <td>{formatMoney(subTotal)}</td>
                  </tr>

                  <tr>
                    <td>Shipping</td>
                    <td>{formatMoney(totalShipping)}</td>
                  </tr>

                  <tr className="totals-table-last-item">
                    <td>Sales Tax</td>
                    <td>{formatMoney(totalSalesTax)}</td>
                  </tr>

                  <tr>
                    <td>Estimated Total</td>
                    <td>{formatMoney(totalAmount)}</td>
                  </tr>
                </tbody>
              </StyledTotalsTable>
            </div>

            <div className="cart-close">
              <CheckoutCart
                totalQuantity={totalQuantity}
                totalAmount={totalAmount}
                email={!!me ? me.email : ""}
                image_url={imgUrl}
                disabled={!me || !totalQuantity || !totalAmount}
              >
                Checkout
              </CheckoutCart>
            </div>
          </footer>
        </StyledCart>
      )
    }}
  </Query>
);


export default Cart;
