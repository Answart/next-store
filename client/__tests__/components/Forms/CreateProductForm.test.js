import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import { CreateProductForm } from '../../../components/Forms';
import {
  mockImage, mockImageVariables, mockProduct,
  createImageMutationMock,
  createProductMutationMock, createProductMutationErrorMock,
} from '../../../lib/test-utils/mocks';

const successMocks = [
  createImageMutationMock(),
  { ...createProductMutationMock },
];
const errorMocks = [
  createImageMutationMock(),
  { ...createProductMutationErrorMock },
];
// mock the global fetch API (cloudinary)
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    public_id: mockImage.cloudinary_id,
    original_filename: mockImage.name,
    secure_url: mockImage.image_url,
    delete_token: 'delete tokeeeen',
    eager: [{
      height: mockImage.height,
      width: mockImage.width,
      transformation: mockImage.transformation,
      secure_url: mockImage.large_image_url,
    }]
  }),
});


describe('<CreateProductForm />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={successMocks} addTypename={false}>
        <CreateProductForm />
      </MockedProvider>
    );
  });
  afterEach(() => jest.clearAllMocks());
  afterAll(() => {
    wrapper.unmount();
    global.fetch.mockReset();
  });

  it('renders and matches snapshot', async () => {
    const form = wrapper.find('form[data-test="form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('handles state updating', async () => {
    wrapper.find('#title').simulate('change', { target: {
      value: mockProduct.title, name: 'title'
    }});
    wrapper.find('#department').simulate('change', { target: {
      value: mockProduct.department, name: 'department', type: 'select'
    }});
    wrapper.find('#category').simulate('change', { target: {
      value: mockProduct.category, name: 'category', type: 'select'
    }});
    wrapper.find('#description').simulate('change', { target: {
      value: mockProduct.description, name: 'description'
    }});
    wrapper.find('#brand').simulate('change', { target: {
      value: mockProduct.brand, name: 'brand'
    }});
    wrapper.find('#offline').simulate('change', { target: {
      value: mockProduct.online, name: 'offline', type: 'radio'
    }});
    wrapper.update();
    expect(wrapper.find('CreateProductForm').instance().state).toMatchObject({
      title: mockProduct.title,
      department: mockProduct.department,
      category: mockProduct.category,
      description: mockProduct.description,
      brand: mockProduct.brand,
      online: mockProduct.online,
    });
  });

  it('image upload updates image in state', async () => {
    wrapper.find('input[type="file"]').simulate('change', { target: {
      files: [mockImage.image_url], name: 'image', type: 'file'
    }});
    await wait(50);
    wrapper.update();
    expect(wrapper.find('CreateProductForm').instance().state.image).toMatchObject({
      ...mockImageVariables,
      delete_token: 'delete tokeeeen'
    });
  });

  it('routes to product/selections after successful product creation', async () => {
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await wait(100);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/product/selections',
      query: { id: mockProduct.id },
    });
  });

  it('renders DisplayMessage component w/error message when submitted with invalid/insufficient info', async () => {
    console.error = jest.fn();
    const wrapper2 = mount(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <CreateProductForm />
      </MockedProvider>
    );
    wrapper2.find('input[type="file"]').simulate('change', { target: {
      files: [mockImage.image_url], name: 'image', type: 'file'
    }});
    await wait(50);
    expect(global.fetch).toHaveBeenCalled();
    Router.router = { push: jest.fn() };
    wrapper2.find('form').simulate('submit');
    await wait(100);
    wrapper2.update();
    expect(wrapper2.find('DisplayMessage').text()).toContain('Hold up! ack!');
    expect(Router.router.push).not.toHaveBeenCalled();
    wrapper2.unmount();
  });
});
