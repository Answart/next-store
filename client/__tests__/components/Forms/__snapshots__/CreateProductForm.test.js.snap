// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProductForm /> renders and matches snapshot 1`] = `null`;
